### PR TITLE
feat: Add total lattice gradient and per-axis periodicity masking for variable-cell relaxation

### LINKED ISSUE #160: Relaxation of lattice should consider particle positions
### What broke?

Currently, we only consider the partial gradient of the lattice vectors but we should optimize using the total gradient.

### How to reproduce

```python

```

### Expected behavior

_No response_

### Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by n-gao**
### Overview

This PR extends structure relaxation to support variable-cell optimisation and adds correct handling of per-axis periodicity throughout the stress and lattice gradient machinery.

### `MappedPotential` API Refactor

`MappedPotential` previously accepted separate `gradient_map` and `hessian_map` callables. These have been replaced with a single `mapping` callable that receives a `MappedPotentialInput` named tuple containing both the current `state` and the full `PotentialOut`. This gives mapping functions access to the state (e.g. positions, cell) when computing derived quantities, which is required for the total lattice gradient.

### Total Lattice Gradient

A new `total_lattice_gradient` function in `kups/observables/stress.py` computes the total derivative `∂E/∂h` at fixed fractional coordinates, adding the position-virial coupling term `h⁻ᵀ·Σᵢ rᵢ⊗∂E/∂rᵢ` to the partial gradient reported by the potential. Non-periodic axes are masked out so that vacuum/bounding-box basis vectors do not accumulate spurious coupling.

### Variable-Cell Relaxation

The relaxation propagator now applies `total_lattice_gradient` when mapping potential output, replacing the previous pass-through of the raw cell gradient. The convergence check in `run_relax` has been extended to include the maximum absolute cell gradient alongside the maximum force, using a single `force_tolerance` threshold for both.

### Per-Axis Periodicity in Stress

`stress_via_virial_theorem` and `_molecular_stress_via_virial_theorem` now accept `Cell[AnyPeriodicity]` instead of `Cell[Periodic3D]`. A `_periodic_mask` helper zeros stress components on non-periodic axes, so isolated clusters produce zero stress and slabs retain only their in-plane block. `IsVirialSystems` and `RelaxSystems` have been updated accordingly.

### Tests

- `TestCellRelaxation` verifies that `optimize_cell=True` moves the lattice vectors while `optimize_cell=False` holds them fixed.
- `TestTotalLatticeGradient` validates the coupling term against `jax.grad` through fixed-fractional-coordinate energy evaluation, checks that zero forces reduce to the partial gradient, and confirms that non-periodic axes drop the coupling entirely.
- `TestStressViaVirialTheorem` gains cases for a fully non-periodic (vacuum) cell producing zero stress and a slab geometry zeroing the non-periodic row and column while preserving the in-plane block.  
  
Closes #160

### CODE REVIEW

**Comment by n-gao** [comment]
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#215](https://github.com/cusp-ai-oss/kUPS/pull/215) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by jonkhler — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Jun 24, 1:04 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178).
